File: pycompiler/AST.py
from typing import Union
from collections import namedtuple

Boolean = namedtuple('Boolean', 'value')
Nil = namedtuple('Nil', '')
Symbol = namedtuple('Symbol', 'value')
Number = namedtuple('Number', 'value')
String = namedtuple('String', 'value')
Chain = namedtuple('Chain', 'values')
ByteArray = namedtuple('ByteArray', 'values')
Block = namedtuple('Block', 'args statements')
Array = namedtuple('Array', 'elements')
Identifier = namedtuple('Identifier', 'name')
Value = Union[Identifier, Array, Block, ByteArray, Chain, String, Number, Symbol]

UnaryMessage = namedtuple('UnaryMessage', 'name')
BinaryMessage = namedtuple('BinaryMessage', 'name argument')
KeywordMessage = namedtuple('KeywordMessage', 'name argdict')
Message = Union[KeywordMessage, BinaryMessage, UnaryMessage]

UnaryExpression = namedtuple('UnaryExpression', 'receiver messages')
BinaryExpression = namedtuple('BinaryExpression', 'receiver messages')
KeywordExpression = namedtuple('KeywordExpression', 'receiver message')
Expression = namedtuple('Expression', 'value cascade')

Signature = namedtuple('Signature', 'name arguments')
Requirement = namedtuple('Requirement', 'signature')
Method = namedtuple('Method', 'static signature statements')
Field = namedtuple('Field', 'static name value')
Member = Union[Field, Method, Requirement]

Pragma = namedtuple('Pragma', 'value')
Answer = namedtuple('Answer', 'value')
TraitDef = namedtuple('TraitDef', 'name parent_trait members')
ClassDef = namedtuple('ClassDef', 'name parent_class traits members')
Assignment = namedtuple('Assignment', 'name value')
TempDecl = namedtuple('TempDecl', 'name value')
Statement = Union[Assignment, ClassDef, TraitDef, Answer, Pragma]

Program = namedtuple('Program', 'statements')

AST = Union[Program, Statement, Member, Expression, Message, Value]
BlockContext = Union[Program, Method, Block]

def inUnion(_type, union):
    print("In Union called with:", _type)
    return _type in union.__args__

File: JavaCompiler/src/com/smallos/AST.java
package com.smallos;
import java.util.List;
import java.util.Map;

public interface AST {
    public static interface Node {}
    
    public static interface Value extends Expr {}
    public static record Bool(boolean value) implements Value;
    public static record Nil() implements Value;
    public static record Symbol(String value) implements Value;
    public static record Num(Number value) implements Value;
    public static record Str(String value) implements Value;
    public static record ByteBlock(byte[] value) implements Value;
    public static record Identifier(String name) implements Value;
    public static record Chain(List<Expr> values) implements Value;
    public static record Block(List<Identifier> args, List<Stmt> statements) implements Value;
    public static record _Array(List<Expr> values) implements Value;
    
    public static interface Message extends Node {}
    public static record UnaryMessage(String name) implements Message;
    public static record BinaryMessage(String name, UnaryExpression argument) implements Message;
    public static record KeywordMessage(String name, Map<String, BinaryMessage> arguments) implements Message;
    
    public static interface Expr extends Node {}
    public static record UnaryExpression(Value receiver, List<UnaryMessage> messages) implements Expr;
    public static record BinaryExpression(UnaryExpression receiver, List<BinaryMessage> message) implements Expr;
    public static record KeywordExpression(BinaryExpression receiver, KeywordMessage message) implements Expr;
    public static record Cascade(KeywordExpression primary, Message messages) implements Expr;
    
    public static interface Member extends Node {}
    public static record Requirement(Signature signature) implements Member;
    public static record Method(boolean isStatic, Signature signature, List<Stmt> statements) implements Member;
    public static record Field(boolean isStatic, String name, Expr value) implements Member;
    
    public static interface Signature extends Node {}
    public static record UnarySignature(String name) implements Signature;
    public static record BinarySignature(String name, Identifier argument) implements Signature;
    public static record KeywordSignature(String name, Map<String, Identifier> arguments) implements Signature;
    
    public static interface Stmt extends Node {}
    public static record Pragma(Message value) implements Stmt, Member;
    public static record Answer(Expr value) implements Stmt;
    public static record TraitDef(String name, Identifier parentTrait, List<Member> members) implements Stmt;
    public static record ClassDef(String name, Identifier parentClass, List<Identifier> traits, List<Member> members) implements Stmt;
    public static record Assignment(Identifier name, Expr value) implements Stmt;
    public static record TempDecl(Identifier name, Expr value) implements Stmt;
    
    public static record Stmt(List<Stmt> statements) extends Node;
}
